File: kindle/kpm/skibidi-display/touch-exit.lua
local ffi = require("ffi")

ffi.cdef[[
struct input_event {
  long tv_sec;
  long tv_usec;
  unsigned short type;
  unsigned short code;
  int value;
};
int open(const char *pathname, int flags);
int close(int fd);
long read(int fd, void *buf, unsigned long count);
]]

local C = ffi.C
local DEVICE = "/dev/input/event2"
local STOP = "/mnt/us/kmc/kpm/packages/skibidi-display/stop.sh"
local LOG = "/mnt/us/skibidi-touch-exit.log"

local EV_SYN, EV_ABS = 0, 3
local SYN_REPORT = 0
local ABS_MT_POSITION_X, ABS_MT_POSITION_Y, ABS_MT_TRACKING_ID = 53, 54, 57

local function log(msg)
  local f = io.open(LOG, "a")
  if f then
    f:write(os.date("%Y-%m-%d %H:%M:%S "), msg, "\n")
    f:close()
  end
end

local fd = C.open(DEVICE, 0)
if fd < 0 then
  log("ERROR cannot open " .. DEVICE)
  os.exit(1)
end

local ev = ffi.new("struct input_event")
local size = ffi.sizeof(ev)
local x, y
local down = false
local down_at = 0
local min_x, max_x, min_y, max_y

local function observe(v, lo, hi)
  if lo == nil or v < lo then lo = v end
  if hi == nil or v > hi then hi = v end
  return lo, hi
end

local function is_corner(px, py)
  if not (px and py and min_x and max_x and min_y and max_y) then return false end
  local xr, yr = max_x - min_x, max_y - min_y
  if xr < 100 or yr < 100 then return false end
  -- For the first hardware proof, accept a long hold in any corner because
  -- Goodix axis orientation may be mirrored/rotated relative to the screen.
  local xedge = px <= min_x + xr * 0.28 or px >= min_x + xr * 0.72
  local yedge = py <= min_y + yr * 0.22 or py >= min_y + yr * 0.78
  return xedge and yedge
end

log("Lua touch watcher starting; event_size=" .. tostring(size))

while true do
  local n = C.read(fd, ev, size)
  if n == size then
    local etype, code, value = tonumber(ev.type), tonumber(ev.code), tonumber(ev.value)
    if etype == EV_ABS then
      if code == ABS_MT_POSITION_X then
        x = value
        min_x, max_x = observe(value, min_x, max_x)
      elseif code == ABS_MT_POSITION_Y then
        y = value
        min_y, max_y = observe(value, min_y, max_y)
      elseif code == ABS_MT_TRACKING_ID then
        if value >= 0 and not down then
          down = true
          down_at = os.time()
          log("DOWN id=" .. value .. " x=" .. tostring(x) .. " y=" .. tostring(y))
        elseif value < 0 and down then
          local held = os.time() - down_at
          log("UP x=" .. tostring(x) .. " y=" .. tostring(y) ..
              " held=" .. tostring(held) ..
              " range=" .. tostring(min_x) .. "," .. tostring(min_y) ..
              ".." .. tostring(max_x) .. "," .. tostring(max_y))
          if held >= 2 and is_corner(x, y) then
            log("EXIT gesture accepted")
            C.close(fd)
            os.execute("/bin/sh " .. STOP .. " >/tmp/skibidi-stop.log 2>&1")
            os.exit(0)
          end
          down = false
        end
      end
    elseif etype == EV_SYN and code == SYN_REPORT then
      -- Coordinates are committed here; no action required until release.
    end
  elseif n < 0 then
    log("ERROR read failed")
    C.close(fd)
    os.exit(1)
  end
end

File: kindle/kpm/skibidi-display/touch-exit.py
#!/usr/bin/env python3
import os, struct, time

DEVICE = "/dev/input/event2"
STOP = "/mnt/us/kmc/kpm/packages/skibidi-display/stop.sh"
LOG = "/mnt/us/skibidi-touch-exit.log"

EV_SYN, EV_ABS = 0, 3
SYN_REPORT = 0
ABS_MT_POSITION_X, ABS_MT_POSITION_Y, ABS_MT_TRACKING_ID = 53, 54, 57

# The on-screen EXIT target is the top-right corner. We deliberately define it
# as a percentage of the digitizer's observed coordinate range, so it works
# without hard-coding this Kindle panel's absolute max values.
x = y = None
min_x = min_y = None
max_x = max_y = None
tracking = False
started = None
start_x = start_y = None

def log(msg):
    try:
        with open(LOG, "a") as f:
            f.write("%s %s\n" % (time.strftime("%Y-%m-%d %H:%M:%S"), msg))
    except Exception:
        pass

def in_exit(px, py):
    if None in (px, py, min_x, min_y, max_x, max_y):
        return False
    xr = max_x - min_x
    yr = max_y - min_y
    if xr < 100 or yr < 100:
        return False
    # Top-right 28% x 22% of the observed panel range. Because some panels
    # report inverted axes, accept any corner for the first hardware proof;
    # the visible EXIT target is top-right and we'll tighten orientation after.
    near_x_edge = px >= min_x + xr * 0.72 or px <= min_x + xr * 0.28
    near_y_edge = py >= min_y + yr * 0.78 or py <= min_y + yr * 0.22
    return near_x_edge and near_y_edge

log("touch exit watcher starting on %s" % DEVICE)

fmt = "llHHi"
size = struct.calcsize(fmt)

try:
    fd = os.open(DEVICE, os.O_RDONLY)
    while True:
        raw = os.read(fd, size)
        if len(raw) != size:
            continue
        sec, usec, etype, code, value = struct.unpack(fmt, raw)

        if etype == EV_ABS:
            if code == ABS_MT_POSITION_X:
                x = value
                min_x = value if min_x is None else min(min_x, value)
                max_x = value if max_x is None else max(max_x, value)
            elif code == ABS_MT_POSITION_Y:
                y = value
                min_y = value if min_y is None else min(min_y, value)
                max_y = value if max_y is None else max(max_y, value)
            elif code == ABS_MT_TRACKING_ID:
                if value >= 0 and not tracking:
                    tracking = True
                    started = time.time()
                    start_x, start_y = x, y
                elif value < 0 and tracking:
                    duration = time.time() - started if started else 0
                    ex, ey = x, y
                    log("release x=%s y=%s held=%.2f range=%s,%s..%s,%s" %
                        (ex, ey, duration, min_x, min_y, max_x, max_y))
                    if duration >= 1.5 and in_exit(ex, ey):
                        log("EXIT gesture accepted")
                        os.system("/bin/sh %s >/tmp/skibidi-stop.log 2>&1" % STOP)
                        break
                    tracking = False
                    started = None
        elif etype == EV_SYN and code == SYN_REPORT:
            if tracking and start_x is None and x is not None and y is not None:
                start_x, start_y = x, y
except Exception as exc:
    log("watcher failed: %r" % (exc,))
